types/trace.go

```diff
@@ -0,0 +1,102 @@
+package types
+
+import (
+	"encoding/json"
+	"time"
+)
+
+// TraceSession model.
+type TraceSession struct {
+	ID         string        `json:"id" yaml:"id"`
+	PipelineID string        `json:"pipelineID" yaml:"pipelineID"`
+	Plugins    []string      `json:"plugins" yaml:"plugins"`
+	Lifespan   time.Duration `json:"lifespan" yaml:"lifespan"`
+	CreatedAt  time.Time     `json:"createdAt" yaml:"createdAt"`
+	UpdatedAt  time.Time     `json:"updatedAt" yaml:"updatedAt"`
+}
+
+// Active tells whether a session is still within its lifespan.
+func (ts TraceSession) Active() bool {
+	return ts.CreatedAt.Add(ts.Lifespan).After(time.Now())
+}
+
+// CreateTraceSession request payload for creating a new trace session.
+type CreateTraceSession struct {
+	Plugins  []string      `json:"tracePlugins"`
+	Lifespan time.Duration `json:"traceLifespan"`
+}
+
+// TraceSessionsParams request payload for querying trace sessions.
+type TraceSessionsParams struct {
+	Last   *uint64
+	Before *string
+}
+
+// TraceSessions paginated list.
+type TraceSessions struct {
+	Items     []TraceSession
+	EndCursor *string
+}
+
+// UpdateTraceSession request payload for updating a trace session.
+type UpdateTraceSession struct {
+	Plugins  *[]string      `json:"plugins"`
+	Lifespan *time.Duration `json:"lifespan"`
+}
+
+// TraceRecord model.
+type TraceRecord struct {
+	ID        string    `json:"id" yaml:"id"`
+	SessionID string    `json:"sessionID" yaml:"sessionID"`
+	CreatedAt time.Time `json:"createdAt" yaml:"createdAt"`
+
+	// fluent-bit data from here on.
+	Kind           TraceRecordKind `json:"type" yaml:"type"`
+	TraceID        string          `json:"traceID" yaml:"traceID"`
+	StartTime      time.Time       `json:"start_time" yaml:"start_time"`
+	EndTime        time.Time       `json:"end_time" yaml:"end_time"`
+	InputInstance  string          `json:"input_instance" yaml:"input_instance"`
+	FilterInstance string          `json:"filter_instance" yaml:"filter_instance"`
+	OutputInstance string          `json:"outputInstance" yaml:"outputInstance"`
+	ReturnCode     int             `json:"return_code" yaml:"return_code"`
+	// Each record is a JSON object,
+	// warranted to have a flb_time `timestamp` field.
+	Records []json.RawMessage `json:"records" yaml:"records"`
+}
+
+// TraceRecordKind enum.
+type TraceRecordKind string
+
+const (
+	TraceRecordKindInput     TraceRecordKind = "input"
+	TraceRecordKindFilter    TraceRecordKind = "filter"
+	TraceRecordKindPreOutput TraceRecordKind = "pre_output"
+	TraceRecordKindOutput    TraceRecordKind = "output"
+)
+
+// CreateTraceRecord request payload for creating a new trace record.
+type CreateTraceRecord struct {
```

Comment:
@pwhelan what do you think about this struct?
This is the data the agent is gonna send to cloud.

The endpoint is gonna be something like:
```http
POST /v1/pipelines/{pipelineID}/trace_session/records
```

So it pushes the record to the current active trace session on the pipeline.

Reply:
Notice that the endpoint is "trace_session" in singular since it's not the entire resource but just the current active session.

Reply:
> @pwhelan what do you think about this struct? This is the data the agent is gonna send to cloud.

Fluent-bit should push the records straight to the cloud URL. We do not need an actual agent to transfer from fluent-bit to the cloud. Would it work better if we changed `(filter|input|output)_instance` field name to `plugin_instance`? Should I also send the alias that way it is easier to link up the plugin with its trace records in a reliable way?

> The endpoint is gonna be something like:
> 
> ```httpspec
> POST /v1/pipelines/{pipelineID}/trace_session/records
> ```

LGTM.

Reply:
Sure. Ping me if you do the change.

Reply:
This change has been applied.